fix(nrh): fixes for ESP contact syncing w/ NRH, newsletter signups

dkoo opened the pull request:
### All Submissions:

* [x] Have you followed the [Newspack Contributing guideline](https://github.com/Automattic/newspack-plugin/blob/trunk/.github/CONTRIBUTING.md)?
* [x] Does your code follow the [WordPress' coding standards](https://make.wordpress.org/core/handbook/best-practices/coding-standards/) and [VIP Go coding standards](https://vip.wordpress.com/documentation/vip-go/code-review-blockers-warnings-notices/)?
* [x] Have you checked to ensure there aren't other open [Pull Requests](../../pulls) for the same update/change?

<!-- Mark completed items with an [x] -->

<!-- You can erase any parts of this template not applicable to your Pull Request. -->

### Changes proposed in this Pull Request:

<!-- Describe the changes made to this Pull Request, and the reason for such changes. -->

It's possible to enable Reader Activation contact syncing while also using NRH instead of WooCommerce as a reader revenue platform, but this is currently broken.

Handles ESP contact syncing when using NRH as a reader revenue platform, and/or WooCommerce Subscriptions is not available. WooCommerce itself is still a soft requirement for ESP syncing when using NRH, as we rely on its `WC_Customer` class for fetching basic user meta fields.

Also fixes two scenarios where some contact metadata is lost before getting synced:

- Upon registering/subscribing to newsletters via the Registration and Newsletter Subscription `newspack_newsletters_contact_subscribed` action triggers a `newsletter_subscribed` data event that contains the full contact metadata in the payload, but while the handler for this data event triggers a contact sync, it doesn't include the full metadata in the sync. This PR fixes the issue by including all metadata in the sync.
- When queued data events include certain metadata that's not included in subsequent events in the queue, that metadata is lost. This PR fixes the issue by merging contact metadata from all queued data events before syncing them.

Closes `1200550061930446/1209469667970736`.

### How to test the changes in this Pull Request:

#### Testing synced metadata when registering via signup blocks

1. With Reader Activation syncing active with all contact fields enabled and connected to an ESP, enable the Newspack reader revenue platform.
2. In new reader sessions, sign up for newsletters via the Registration and Newsletter Subscription Form blocks.
3. On `release`, observe that some contact metadata fields are not synced to the ESP (NP_Registration Page, any Signup UTM fields).
4. Check out this branch and repeat, and confirm that all registration-related fields are synced.

#### Testing contact syncing with NRH

1. On a site with Reader Activation syncing active and connected to an ESP, set your reader revenue platform to NRH and deactivate the WooCommerce Subscriptions plugin.
5. On `release`, as a reader, sign up for newsletters via the Newsletter Subscription Form and Reader Registration blocks on a page that contains UTM parameters in the URL (e.g. `?utm_medium=test&utm_campaign=NRH&utm_content=NRH%20content`. Observe that the contact sync to the connected ESP is logged, but without most of the RAS contact metadata fields. Also observe a fatal error in PHP: `Fatal error: Uncaught Error: Call to undefined function wcs_get_users_subscriptions()` 
6. Check out this branch and https://github.com/Automattic/newspack-newsletters/pull/1763. As an admin, navigate to **Newspack > Reader Activation > Advanced Settings > Metadata field settings**. Confirm that only basic metadata fields are available (as payment fields assume that the reader revenue platform is WooCommerce):

<img width="999" alt="Screenshot 2025-02-24 at 12 16 23 PM" src="https://github.com/user-attachments/assets/a1a061dc-c839-48e5-a752-5a027d63bdf0" />

7. Repeat signing up for newsletters with UTM params as in step 2, but this time confirm that all metadata fields are synced to the ESP without the fatal error.
8. Follow testing instructions for https://github.com/Automattic/newspack-newsletters/pull/1648 and ensure that GA4 data is unaffected.

### Other information:

* [x] Have you added an explanation of what your changes do and why you'd like us to include them?
* [ ] Have you written new tests for your changes, as applicable?
* [x] Have you successfully ran tests with your changes locally?

<!-- Mark completed items with an [x] -->

---
- To see the specific tasks where the Asana app for GitHub is being used, see below:
  - https://app.asana.com/0/0/1209571170057170

dkoo committed:
fix(nrh): sync only basic fields; normalize on Newspack_Newsletters_Contacts::upsert

dkoo committed:
fix(nrh): only add Woo Teams meta field if using Woo

dkoo committed:
fix(nrh): better check for avoiding fatal when WC_Customer doesn't exist

dkoo committed:
fix(nrh): more checks for WCS functions

dkoo committed:
Merge branch 'release' into hotfix/ras-nrh-fixes

adekbadek reviewed (commented):
Works accordingly to the testing instructions, but there's an ongoing discussion on the other PR (https://github.com/Automattic/newspack-newsletters/pull/1763#pullrequestreview-2644443781), which should be resolved before this is merged.

dkoo committed:
Merge branch 'release' into hotfix/ras-nrh-fixes

dkoo committed:
Merge branch 'release' into hotfix/ras-nrh-fixes

dkoo committed:
fix(esp-sync): include metadata on newsletter signup and queued syncs

dkoo commented:
@adekbadek @miguelpeixe @leogermani following up from a [comment in the other PR](https://github.com/Automattic/newspack-newsletters/pull/1763#discussion_r1983885984), which is no longer needed after 6b571e9. I updated the PR description here to explain what was happening and how this PR fixes it.

leogermani commented:
Not sure if there's something wrong on my end, but I was not able to confirm most of the things in the testing instructions.

= Testing synced metadata when registering via signup blocks =

Both on `release` and on this branch, I never see "Registration Page" going through. "Registration Method" is always there, on both branches

= Testing contact syncing with NRH =

I see the Fatal error going away and the actual sync happening. Also see the shortened list of fields on the admin. But I don't see neither "Registration Page" or the UTM params going through.... :/

Maybe we can debug it together

dkoo committed:
fix(esp-sync): signup UTM values

dkoo commented:
@leogermani looks like there was an unrelated bug preventing Signup UTM fields from being set. Fixed in 826df46. Also, Registration Method was not affected by any of these bugs. Removed that from testing instructions

> I never see "Registration Page" going through

That doesn't match what I see...on this branch, I do see "Registration Page", but not on `release` or `trunk`. Are you sure the field is enabled in sync settings?

dkoo committed:
Merge branch 'release' into hotfix/ras-nrh-fixes

leogermani commented:
Apparently the last fix fixed it all and I now see the Registration Page. Not sure if this is related to this PR or not, but the contents of this meta looks a bit off. Example:

```
"NP_Registration Page": "https:\/\/leogermani.sa.ngrok.io\/https:\/\/leogermani.sa.ngrok.io\/buy-news\/?utm_medium=test&utm_campaign=NRH&utm_content=NRH+content"
```

dkoo commented:
@leogermani what method did you use to register when you saw the weird double URL? Or does it always happen for you?

leogermani commented:
> @leogermani what method did you use to register when you saw the weird double URL? Or does it always happen for you?

A newsletter subscription block added to a page

dkoo commented:
Aha, I was able to replicate it with the Newsletter Subscription Form block too. I think this is coming from the Newsletters plugin, not this PR

EDIT: it's the `home_url` [here](https://github.com/Automattic/newspack-newsletters/blob/trunk/src/blocks/subscribe/index.php#L478) that's causing the double URL. If the referrer is a full URL instead of a path, then `home_url()` will still append it to the site's URL 😕

leogermani commented:
ok! Let's move this forward and I'll look at this separately

leogermani commented:
https://github.com/Automattic/newspack-newsletters/pull/1777
